otel/otel.go

```diff
@@ -0,0 +1,171 @@
+	ServiceName               string
+	Debug                     bool
+	ExporterType              string
+	ExporterEndpoint          string
+	ExporterTimeout           string
+	ExporterCertificate       string
+	ExporterClientCertificate string
+	ExporterClientKey         string
+	CollectionInterval        time.Duration
+	MetricsReader             sdkmetric.Reader
+	MetricsExporter           sdkmetric.Exporter
+}
+
+// OtelWrapper encapsulates OpenTelemetry MetricProvider and utilities
+type OtelWrapper struct {
+	meterProvider *sdkmetric.MeterProvider
+	globalMeter   metric.Meter
+	config        Config
+}
+
+var (
+	globalWrapper *OtelWrapper
+
+	// singletonShutdown holds the shutdown function once the meter provider is initialized.
+	singletonShutdown func(context.Context) error
+
+	// once ensures that initialization happens only once.
+	globalOnce sync.Once
+
+	// initErr captures any error encountered during initialization.
+	initErr error
+)
+
+// InitSingleton initializes the MeterProvider as a singleton.
+// Subsequent calls to this function will return the same shutdown function and error.
+// The configuration is used only on the first call.
+func InitGlobalWrapper(ctx context.Context, cfg Config) (func(context.Context) error, error) {
+	globalOnce.Do(func() {
+		singletonShutdown, initErr = New(ctx, cfg)
+	})
+	return singletonShutdown, initErr
+}
+
+func New(ctx context.Context, cfg Config) (shutdown func(context.Context) error, err error) {
+	var shutdownFuncs []func(context.Context) error
+
+	// shutdown calls cleanup functions registered via shutdownFuncs.
+	// The errors from the calls are joined.
+	// Each registered cleanup will be invoked once.
+	shutdown = func(ctx context.Context) error {
+		var err error
+		for _, fn := range shutdownFuncs {
+			err = errors.Join(err, fn(ctx))
+		}
+		shutdownFuncs = nil
+		return err
+	}
+
+	// handleErr calls shutdown for cleanup and makes sure that all errors are returned.
+	handleErr := func(inErr error) {
+		err = errors.Join(inErr, shutdown(ctx))
+	}
+
+	// Set up meter provider.
+	meterProvider, err := newMeterProvider(ctx, cfg)
+	if err != nil {
+		handleErr(err)
+		return
+	}
+	shutdownFuncs = append(shutdownFuncs, meterProvider.Shutdown)
+	otelsdk.SetMeterProvider(meterProvider)
+
+	globalWrapper = &OtelWrapper{
+		meterProvider: meterProvider,
+		globalMeter:   meterProvider.Meter(cfg.ServiceName),
+		config:        cfg,
+	}
+
+	return
+}
+
+func newMeterProvider(ctx context.Context, cfg Config) (*sdkmetric.MeterProvider, error) {
+	if cfg.ServiceName == "" {
+		return nil, errors.New("ServiceName is required")
+	}
+	if cfg.MetricsExporter == nil {
+		exporter, err := newMetricsExporter(ctx, cfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load exporter", err)
+		}
+		cfg.MetricsExporter = exporter
+	}
+	if cfg.MetricsReader == nil {
+		cfg.MetricsReader = sdkmetric.NewPeriodicReader(cfg.MetricsExporter, sdkmetric.WithInterval(cfg.CollectionInterval))
+	}
```

Comment:
I'm not sure we want to override MetricReader. I'm not sure we want to use a ManualReader which add more things to do.
The only thing that need to be override is the collection interval if needed.